model/PermissionProvider.php

```diff
@@ -63,29 +67,67 @@ public function getPermissions(User $user, array $resourceIds) {
         return $rights;
     }
 
-    protected function getStateObjects($stateIds) {
+    protected function getStateObjects($stateIds)
+    {
         $wfmodel = $this->getServiceManager()->get(WorkflowModel::SERVICE_ID);
         $stateCache = array();
         foreach ($stateIds as $stateId) {
             if (!empty($stateId)) {
-                $stateCache[$stateId] = $wfmodel->getState($stateId);
+                $state = $wfmodel->getState($stateId);
+                $stateWithRoles = $this->addAtomicRoles($state);
+                $stateCache[$stateId] = $stateWithRoles;
             }
         }
         return $stateCache;
     }
 
     /**
      * (non-PHPdoc)
+     *
      * @see \oat\generis\model\data\PermissionInterface::onResourceCreated()
      */
-    public function onResourceCreated(\core_kernel_classes_Resource $resource) {
+    public function onResourceCreated(\core_kernel_classes_Resource $resource)
+    {
     }
 
     /**
      * (non-PHPdoc)
+     *
      * @see \oat\generis\model\data\permission\PermissionInterface::getSupportedRights()
      */
-    public function getSupportedRights() {
+    public function getSupportedRights()
+    {
         return array('WRITE', 'READ');
     }
-}
\ No newline at end of file
+
+    private function addAtomicRoles(StateObject $state): StateObject
```

Comment:
We shall stick to open/closed, wdyt?

```suggestion
    private function addAtomicRoles(WorkflowModel $state): WorkflowModel
```